app/views/questionnaire.py

```diff
@@ -300,30 +360,46 @@ def post_block_handler(
             questionnaire_store_updater,
             list_item_id,
         )
-        if next_url:
-            return redirect(next_url)
+        if list_action_redirect:
+            return redirect(list_action_redirect)
 
     if _is_end_of_questionnaire(block):
         return submit_answers(schema)
 
     questionnaire_store_updater.update_answers(form)
-    questionnaire_store_updater.add_completed_location()
+
+    section = schema.get_section_for_block_id(block_id)
 
     # recreate routing path if the answers have changed
     if answer_store.is_dirty:
         routing_path = path_finder.routing_path(section)
 
-    current_section_id = schema.get_section_for_block_id(block_id)['id']
-    if path_finder.is_path_complete(routing_path):
-        questionnaire_store_updater.add_completed_section(current_section_id)
+    questionnaire_store_updater.add_completed_location()
+
+    is_path_complete = path_finder.is_path_complete(routing_path)
+
+    if is_path_complete:
+        questionnaire_store_updater.update_section_status(CompletionStatus.COMPLETED)
     else:
-        questionnaire_store_updater.remove_completed_section(current_section_id)
+        questionnaire_store_updater.update_section_status(CompletionStatus.IN_PROGRESS)
 
     questionnaire_store_updater.save()
 
-    next_location = router.get_next_location(current_location, routing_path)
+    next_location_url = router.get_next_location_url(current_location, routing_path)
+
+    return redirect(next_location_url)
+
+
+def _perform_form_action(
```

Comment:
`form_action` is too generic here as it's only dealing with signing out, `_handle_sign_out` possibly?